File: app/src/main/java/bassamalim/hidaya/activities/BooksActivity.java
package bassamalim.hidaya.activities;

import android.app.DownloadManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.cardview.widget.CardView;
import androidx.preference.PreferenceManager;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.gson.Gson;

import java.io.File;
import java.util.Objects;

import bassamalim.hidaya.R;
import bassamalim.hidaya.databinding.ActivityBooksBinding;
import bassamalim.hidaya.dialogs.TutorialDialog;
import bassamalim.hidaya.models.Book;
import bassamalim.hidaya.other.Global;
import bassamalim.hidaya.other.Utils;

public class BooksActivity extends AppCompatActivity {

    private ActivityBooksBinding binding;
    private Gson gson;
    private int numOfBooks;
    private final String prefix = "/Books/";
    private boolean[] downloaded;
    private CardView[] cards;
    private ImageButton[] downloadBtns;
    private Book.BookInfo[] infoArr;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        binding = ActivityBooksBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        gson = new Gson();

        numOfBooks = getResources().getStringArray(R.array.books_titles).length;
        downloaded = new boolean[numOfBooks];

        checkFirstTime();

        checkDownloaded();

        getInfoArr();

        initViews();

        setListeners();
    }

    private void checkFirstTime() {
        String key = "is_first_time_in_books_activity";
        if (PreferenceManager.getDefaultSharedPreferences(this)
                .getBoolean(key, true))
            new TutorialDialog(this,
                    getString(R.string.books_activity_tips), key)
                    .show(this.getSupportFragmentManager(), TutorialDialog.TAG);
    }

    private void checkDownloaded() {
        File dir = new File(getExternalFilesDir(null) + prefix);

        if (!dir.exists())
            return;

        File[] files = dir.listFiles();

        for (int i = 0; i < Objects.requireNonNull(files).length; i++) {
            File file = files[i];

            String name = file.getName();
            String n = name.substring(0, name.length()-5);
            try {
                int num = Integer.parseInt(n);
                downloaded[num] = true;
            }
            catch (NumberFormatException ignored) {}
        }
    }

    private void getInfoArr() {
        infoArr = new Book.BookInfo[numOfBooks];
        String[] bookTitles = getResources().getStringArray(R.array.books_titles);
        String[] bookAuthors = getResources().getStringArray(R.array.books_authors);
        for (int i = 0; i < numOfBooks; i++) {
            if (downloaded[i]) {
                String path = getExternalFilesDir(null) + "/Books/" + i + ".json";
                String jsonStr = Utils.getJsonFromDownloads(path);
                Book book = gson.fromJson(jsonStr, Book.class);
                infoArr[i] = new Book.BookInfo(i, bookTitles[i], bookAuthors[i]);
            }
        }
    }

    private void initViews() {
        cards = new CardView[]{binding.bokhariCard, binding.muslimCard};
        TextView[] titleTvs = new TextView[]{binding.bokhariTitleTv, binding.muslimTitleTv};
        downloadBtns = new ImageButton[]{binding.bokhariDownloadBtn, binding.muslimDownloadBtn};

        for (int i = 0; i < numOfBooks; i++) {
            if (downloaded[i])
                titleTvs[i].setText(infoArr[i].getBookTitle());
        }

        for (int i = 0; i < numOfBooks; i++)
            updateUI(i, downloaded[i]);
    }

    private void setListeners() {
        for (int i = 0; i < numOfBooks; i++) {
            int finalI = i;
            cards[i].setOnClickListener(v -> {
                if (downloaded(finalI)) {
                    if (downloading(finalI))
                        Toast.makeText(this, getString(R.string.wait_for_download),
                                Toast.LENGTH_SHORT).show();
                    else {
                        Intent intent = new Intent(this,
                                BooksChaptersCollectionActivity.class);
                        intent.putExtra("book_id", finalI);
                        intent.putExtra("book_title",
                                getResources().getStringArray(R.array.books_titles)[finalI]);
                        startActivity(intent);
                    }
                }
                else {
                    getLinkAndRequestDownload(finalI);
                    updateUI(finalI, true);
                }
            });
            downloadBtns[i].setOnClickListener(v -> {
                if (downloaded(finalI)) {
                    if (downloading(finalI))
                        Toast.makeText(this, getString(R.string.wait_for_download),
                                Toast.LENGTH_SHORT).show();
                    else {
                        Utils.deleteFile(this, prefix + finalI + ".json");
                        updateUI(finalI, false);
                    }
                }
                else {
                    getLinkAndRequestDownload(finalI);
                    updateUI(finalI, true);
                }
            });
        }

        binding.fab.setOnClickListener(v -> {
            Intent intent = new Intent(this, BookSearcher.class);
            startActivity(intent);
        });
    }

    private boolean downloaded(int id) {
        File dir = new File(getExternalFilesDir(null) + prefix);

        if (!dir.exists())
            return false;

        File[] files = dir.listFiles();

        for (int i = 0; i < Objects.requireNonNull(files).length; i++) {
            File file = files[i];

            String name = file.getName();
            String n = name.substring(0, name.length()-5);
            try {
                int num = Integer.parseInt(n);
                if (num == id)
                    return true;
            }
            catch (NumberFormatException ignored) {}
        }
        return false;
    }

    private boolean downloading(int id) {
        Log.d(Global.TAG, "here");
        String path = getExternalFilesDir(null) + "/Books/" + id  + ".json";
        String jsonStr = Utils.getJsonFromDownloads(path);
        try {
            gson.fromJson(jsonStr, Book.class);
            return false;
        } catch (Exception e) {
            return true;
        }
    }

    private void updateUI(int id, boolean downloaded) {
        if (downloaded)
            downloadBtns[id].setImageDrawable(AppCompatResources.getDrawable(
                    this, R.drawable.ic_downloaded));
        else
            downloadBtns[id].setImageDrawable(AppCompatResources.getDrawable(
                    this, R.drawable.ic_download));
    }

    private void getLinkAndRequestDownload(int id) {
        String[] links = new String[2];
        FirebaseRemoteConfig remoteConfig = FirebaseRemoteConfig.getInstance();
        remoteConfig.fetchAndActivate().addOnCompleteListener(this, task -> {
            if (task.isSuccessful()) {
                links[0] = remoteConfig.getString("sahih_albokhari_url");
                links[1] = remoteConfig.getString("sahih_muslim_url");

                Log.d(Global.TAG, "Config params updated");
                Log.d(Global.TAG, "Sahih Albokhari URL: " + links[0]);
                Log.d(Global.TAG, "Sahih Muslim URL: " + links[1]);

                download(id, links[id]);
            }
            else
                Log.d(Global.TAG, "Fetch failed");
        });
    }

    private void download(int id, String link) {
        Log.d(Global.TAG, link);

        DownloadManager downloadManager = (DownloadManager)
                getSystemService(Context.DOWNLOAD_SERVICE);
        Uri uri = Uri.parse(link);
        DownloadManager.Request request = new DownloadManager.Request(uri);
         String title = String.format(getString(R.string.book_downloading),
                getResources().getStringArray(R.array.books_titles)[id]);
        request.setTitle(title);
        request.setVisibleInDownloadsUi(true);
        Utils.createDir(this, prefix);
        request.setDestinationInExternalFilesDir(this, prefix, id + ".json");
        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE);
        downloadManager.enqueue(request);
    }

}

File: app/src/main/java/bassamalim/hidaya/activities/BooksActivity.kt
package bassamalim.hidaya.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import bassamalim.hidaya.R
import bassamalim.hidaya.database.dbs.BooksDB
import bassamalim.hidaya.enum.DownloadState
import bassamalim.hidaya.models.Book
import bassamalim.hidaya.other.Global
import bassamalim.hidaya.ui.components.*
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import bassamalim.hidaya.utils.DBUtils
import bassamalim.hidaya.utils.FileUtils
import com.google.firebase.ktx.Firebase
import com.google.firebase.storage.ktx.storage
import com.google.gson.Gson
import java.io.File

class BooksActivity : AppCompatActivity() {

    private lateinit var language: String
    private val prefix = "/Books/"
    private val gson = Gson()
    private lateinit var books: List<BooksDB>
    private val downloadStates = mutableStateListOf<DownloadState>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        language = ActivityUtils.myOnActivityCreated(this)[1]

        books = DBUtils.getDB(this).booksDao().getAll()

        setContent {
            AppTheme {
                UI()
            }
        }
    }

    override fun onResume() {
        super.onResume()

        checkDownloads()
    }

    private fun checkDownloads() {
        downloadStates.clear()
        for (book in books) {
            downloadStates.add(
                if (isDownloaded(book.id)) {
                    if (downloading(book.id)) DownloadState.Downloading
                    else DownloadState.Downloaded
                }
                else DownloadState.NotDownloaded
            )
        }
    }

    private fun isDownloaded(id: Int): Boolean {
        val dir = File(getExternalFilesDir(null).toString() + prefix)

        if (!dir.exists()) return false

        val files = dir.listFiles()
        for (i in 0 until files!!.size) {
            val name = files[i].name
            val n = name.substring(0, name.length - 5)
            try {
                val num = n.toInt()
                if (num == id) return true
            } catch (ignored: NumberFormatException) {}
        }

        return false
    }

    private fun downloading(id: Int): Boolean {
        val path = getExternalFilesDir(null).toString() + "/Books/" + id + ".json"

        val jsonStr = FileUtils.getJsonFromDownloads(path)
        return try {
            gson.fromJson(jsonStr, Book::class.java)
            false
        } catch (e: Exception) {
            true
        }
    }

    private fun download(item: BooksDB) {
        downloadStates[item.id] = DownloadState.Downloading

        val storage = Firebase.storage
        // Create a storage reference from our app
        val storageRef = storage.reference
        // Create a reference with an initial file path and name
        val fileRef = storageRef.child("Books/${item.id}.json")

        FileUtils.createDir(this, prefix)
        val file = File("${getExternalFilesDir(null)}/$prefix/${fileRef.name}")
        file.createNewFile()

        fileRef.getFile(file).addOnSuccessListener {
            Log.i(Global.TAG, "File download succeeded")
            checkDownloads()
        }.addOnFailureListener {
            Log.e(Global.TAG, "File download failed")
        }
    }

    private fun getTitle(item: BooksDB): String {
        return if (language == "en") item.titleEn else item.title
    }

    @Composable
    private fun UI() {
        MyScaffold(
            title = stringResource(R.string.hadeeth_books),
            fab = {
                MyFloatingActionButton(
                    iconId = R.drawable.ic_quran_search,
                    description = stringResource(R.string.search_in_books)
                ) {
                    startActivity(Intent(this, BookSearcher::class.java))
                }
            }
        ) {
            MyLazyColumn(
                Modifier.padding(vertical = 5.dp),
                lazyList = {
                    items(
                        items = books
                    ) { item ->
                        MyBtnSurface(
                            text = item.title,
                            innerVPadding = 15.dp,
                            fontSize = 22.sp,
                            modifier = Modifier.padding(vertical = 2.dp),
                            iconBtn = {
                                MyDownloadBtn(
                                    state = downloadStates[item.id],
                                    path = "$prefix${item.id}.json",
                                    modifier = Modifier.padding(end = 10.dp),
                                    size = 32.dp,
                                    deleted = {
                                        downloadStates[item.id] = DownloadState.NotDownloaded
                                    }
                                ) {
                                    download(item)
                                }
                            }
                        ) {
                            if (downloadStates[item.id] == DownloadState.NotDownloaded)
                                download(item)
                            else if (downloadStates[item.id] == DownloadState.Downloaded) {
                                val intent = Intent(
                                    this@BooksActivity,
                                    BooksChaptersActivity::class.java
                                )
                                intent.putExtra("book_id", item.id)
                                intent.putExtra("book_title", getTitle(item))
                                startActivity(intent)
                            }
                            else FileUtils.showWaitMassage(this@BooksActivity)
                        }
                    }
                }
            )

            TutorialDialog(
                textResId = R.string.books_activity_tips,
                prefKey = "is_first_time_in_books_activity"
            )
        }
    }
}